packages/core/src/VDocument.ts

```diff
@@ -155,4 +153,198 @@ export class VDocument {
+            duplicatedNodes = [...duplicatedNodes, ...duplicatedAncestors];
+
+            // Insert the new list before the first node's list ancestor.
+            // It was split so we need its duplicate as that is where the node is.
+            const newListAncestor = duplicatedAncestors[duplicatedAncestors.length - 1];
+            newListAncestor.before(newList);
+        } else {
+            // Insert the new list before the common ancestor to the first and
+            // last nodes.
+            const commonAncestor = this._commonAncestor(first, last);
+            const reference = first.ancestor(node => node.parent === commonAncestor) || first;
+            commonAncestor.insertBefore(newList, reference);
+        }
+
+        // Move the nodes to the list
+        withMarkers(() => {
+            // Remove the children of nested lists (they will be moved together
+            // with their parent) and convert nested lists to the new type.
+            const nestedLists = nodes.filter(this._isNestedList.bind(this));
+            const nonNestedNodes = nodes.filter(node => {
+                return !node.ancestor(ancestor => nestedLists.includes(ancestor));
+            });
+            nestedLists.forEach(nestedList => ((nestedList as ListNode).listType = type));
+            // Move the nodes.
+            nonNestedNodes.forEach(node => {
+                newList.append(node);
+            });
+            // Remove empty lists.
+            duplicatedNodes.concat(nodes).forEach(node => {
+                if (node.is(ListNode) && !node.hasChildren()) {
+                    node.remove();
+                }
+            });
+        });
+
+        // If the new list is after or before a list of the same type, merge
+        // them (eg: <ol><li>a</li></ol><ol><li>b</li></ol> =>
+        // <ol><li>a</li><li>b</li></ol>).
+        const previousSibling = newList.previousSibling();
+        if (previousSibling && previousSibling.is(ListNode) && previousSibling.listType === type) {
+            newList.children.slice().forEach(child => previousSibling.append(child));
+            newList.remove();
+            newList = previousSibling as ListNode;
+        }
+        const nextSibling = newList.nextSibling();
+        if (nextSibling && nextSibling.is(ListNode) && nextSibling.listType === type) {
+            nextSibling.children.slice().forEach(child => newList.append(child));
+            nextSibling.remove();
+        }
+    }
+    /**
+     * Return true if the given node is a nested `ListNode`, false otherwise.
+     *
+     * @param node
+     */
+    _isNestedList(node: VNode): node is ListNode {
+        return node.is(ListNode) && !!this._listAncestor(node);
+    }
+    /**
+     * Return a node's ancestor that is a `ListNode`, if any.
+     *
+     * @param node
+     */
+    _listAncestor(node: VNode): ListNode {
+        return node.ancestor(ancestor => ancestor.is(ListNode)) as ListNode;
```

Comment:
Just write `return node.ancestor(ListNode)` and typescript will understand that what you want is a `ListNode` ... ;-)